File: Framework/Problem.cs
using System.Collections.Generic;
using System.Linq;

namespace BeetCode.Framework
{
    public abstract class Problem
    {
        public abstract int Number { get; }
        public abstract string Title { get; }
        public abstract string Difficulty { get; }
        public abstract string Description { get; }

        public abstract List<TestCase> GetTestCases();
        public abstract object ExecuteSolution(object[] inputs);

        public void DisplayProblemInfo()
        {
            Console.WriteLine($"Problem {Number}: {Title}");
            Console.WriteLine($"Difficulty: {Difficulty}");
            Console.WriteLine();
            Console.WriteLine(Description);
            Console.WriteLine();
        }

        public bool IsSolved()
        {
            try
            {
                var testCases = GetTestCases();
                if (testCases.Count == 0) return false;

                foreach (var testCase in testCases)
                {
                    var result = ExecuteSolution(testCase.Input);
                    bool success = AreEqual(result, testCase.Expected);
                    if (!success) return false;
                }
                return true;
            }
            catch (NotImplementedException)
            {
                return false;
            }
            catch
            {
                return false;
            }
        }

        protected virtual bool AreEqual(object actual, object expected)
        {
            if (actual == null && expected == null) return true;
            if (actual == null || expected == null) return false;

            // Handle ListNode comparisons
            if (actual is ListNode actualList && expected is ListNode expectedList)
            {
                return AreLinkedListsEqual(actualList, expectedList);
            }

            // Handle TreeNode comparisons
            if (actual is TreeNode actualTree && expected is TreeNode expectedTree)
            {
                return AreTreesEqual(actualTree, expectedTree);
            }

            // Handle arrays
            if (actual is Array actualArray && expected is Array expectedArray)
            {
                return actualArray.Cast<object>().SequenceEqual(expectedArray.Cast<object>());
            }

            return actual.Equals(expected);
        }

        // Helper method for comparing linked lists
        private bool AreLinkedListsEqual(ListNode list1, ListNode list2)
        {
            while (list1 != null && list2 != null)
            {
                if (list1.val != list2.val)
                    return false;
                list1 = list1.next;
                list2 = list2.next;
            }
            
            // Both should be null at the end
            return list1 == null && list2 == null;
        }

        private bool AreTreesEqual(TreeNode node1, TreeNode node2)
        {
            if (node1 == null && node2 == null) return true;
            if (node1 == null || node2 == null) return false;
            return node1.val == node2.val
                && AreTreesEqual(node1.left, node2.left)
                && AreTreesEqual(node1.right, node2.right);
        }

        // Helper method to create a binary tree from level-order array
        protected TreeNode? CreateTree(int?[] values)
        {
            if (values == null || values.Length == 0 || values[0] == null) return null;

            TreeNode root = new TreeNode(values[0].Value);
            Queue<TreeNode> queue = new Queue<TreeNode>();
            queue.Enqueue(root);
            int i = 1;

            while (i < values.Length && queue.Count > 0)
            {
                TreeNode current = queue.Dequeue();

                if (i < values.Length && values[i] != null)
                {
                    current.left = new TreeNode(values[i].Value);
                    queue.Enqueue(current.left);
                }
                i++;

                if (i < values.Length && values[i] != null)
                {
                    current.right = new TreeNode(values[i].Value);
                    queue.Enqueue(current.right);
                }
                i++;
            }

            return root;
        }

        // Helper method to convert linked list to array for display
        protected int[] LinkedListToArray(ListNode head)
        {
            if (head == null) return new int[0];

            List<int> values = new List<int>();
            ListNode current = head;

            while (current != null)
            {
                values.Add(current.val);
                current = current.next;
            }

            return values.ToArray();
        }

        // Helper method to create linked list from array
        protected ListNode? CreateLinkedList(int[] values)
        {
            if (values.Length == 0) return null;

            ListNode head = new ListNode(values[0]);
            ListNode current = head;

            for (int i = 1; i < values.Length; i++)
            {
                current.next = new ListNode(values[i]);
                current = current.next;
            }

            return head;
        }
    }
}

File: TypeScript/framework/problem.ts
import { ListNode, TreeNode, TestCase } from './types';

export abstract class Problem {
	abstract get number(): number;
	abstract get title(): string;
	abstract get difficulty(): string;
	abstract get description(): string;

	abstract getTestCases(): TestCase[];
	abstract executeSolution(inputs: unknown[]): unknown;

	displayProblemInfo(): void {
		console.log(`Problem ${this.number}: ${this.title}`);
		console.log(`Difficulty: ${this.difficulty}`);
		console.log();
		console.log(this.description);
		console.log();
	}

	isSolved(): boolean {
		try {
			const testCases = this.getTestCases();
			if (testCases.length === 0) return false;
			for (const tc of testCases) {
				const result = this.executeSolution(tc.input);
				if (!this.areEqual(result, tc.expected)) return false;
			}
			return true;
		} catch {
			return false;
		}
	}

	areEqual(actual: unknown, expected: unknown): boolean {
		if (actual === null && expected === null) return true;
		if (actual === null || expected === null) return false;

		if (actual instanceof ListNode && expected instanceof ListNode) {
			return linkedListsEqual(actual, expected);
		}
		if (actual instanceof TreeNode && expected instanceof TreeNode) {
			return treesEqual(actual, expected);
		}
		if (Array.isArray(actual) && Array.isArray(expected)) {
			if (actual.length !== expected.length) return false;
			return actual.every((v, i) => this.areEqual(v, (expected as unknown[])[i]));
		}

		return actual === expected;
	}

	protected createLinkedList(values: number[]): ListNode | null {
		if (values.length === 0) return null;
		const head = new ListNode(values[0]);
		let cur = head;
		for (let i = 1; i < values.length; i++) {
			cur.next = new ListNode(values[i]);
			cur = cur.next;
		}
		return head;
	}

	protected linkedListToArray(head: ListNode | null): number[] {
		const out: number[] = [];
		let cur = head;
		while (cur !== null) { out.push(cur.val); cur = cur.next; }
		return out;
	}

	protected createTree(values: (number | null)[]): TreeNode | null {
		if (values.length === 0 || values[0] === null) return null;
		const root = new TreeNode(values[0]);
		const queue: TreeNode[] = [root];
		let i = 1;
		while (i < values.length && queue.length > 0) {
			const cur = queue.shift()!;
			if (i < values.length && values[i] !== null) {
				cur.left = new TreeNode(values[i] as number);
				queue.push(cur.left);
			}
			i++;
			if (i < values.length && values[i] !== null) {
				cur.right = new TreeNode(values[i] as number);
				queue.push(cur.right);
			}
			i++;
		}
		return root;
	}
}

function linkedListsEqual(a: ListNode | null, b: ListNode | null): boolean {
	while (a !== null && b !== null) {
		if (a.val !== b.val) return false;
		a = a.next;
		b = b.next;
	}
	return a === null && b === null;
}

function treesEqual(a: TreeNode | null, b: TreeNode | null): boolean {
	if (a === null && b === null) return true;
	if (a === null || b === null) return false;
	return a.val === b.val && treesEqual(a.left, b.left) && treesEqual(a.right, b.right);
}
